README.md
perfmjs
=======
high performance javascript framework v3.x

perfmjs v3.x Features:
=======
开发目的：对mobile更友好，对开发人员更友好，为未来而设计。

3条军规：

1. 从业务消费点出发做技术

2. 兼顾安全高可靠与高性能

3. 代码“简约易读易改”

v3.0.0新特性：

1. 向上兼容ES6，H5

2. perfmjs-node v3.x

3. 精简代码(删除无用的模块等)

4. 一站式开发工具，文档齐全，可加密打包

5. 统一捕获错误堆栈

6. 充分利用现代浏览器的新特性（如并发）

7. 可定制打包多个版本（mini版、ie6版polyfill、h5版等）


perfmjs V2.x Features
=======
fast by default：高效，易用，易读

核心库(core.js)压缩后只有32k+，还可以优化至更小

实现了浏览器端的AMD（Asynchronous Module Definition）异步模块定义规范，功能类似require.js, 但比require.js更轻量

实现了异步编程模块(Promises/A+)规范

原生态支持面向对象(object-oriented)功能

更高效的插件开发机制

更高效的js和css资源加载功能(底层功能由head.js提供)

更高效的js和css资源文件版本号管理功能

支持高效的数组操作，事件代理，AOP, 状态机, 日志等功能

可测试性，通过了jasmine库的测试

支持所有的主流浏览器，包括：ie5.5+, firefox, chrome, safari, etc

Nodejs版的perfmjs框架请移步：https://github.com/perfmjs/perfmjs-node

可选功能：
=======
jQuery/zepto.js是该框架的可选部分。DOM操作,AJAX,CSS等功能可以使用jQuery及类似框架提供，perfmjs框架在这些功能上不重复发明轮子！


代码质量与构建
=======
WebStorm, SlikSvn(svn命令行客户端), Node.js开发环境, gulp-uglify + gulp-jshint + gulp-jsdoc + 公司的代码规范文档, Jasmine


TODO:
=======
增加资源文件自已更新缓存的功能（不需使用版本号）

去掉namespace(保留base的单例模式)，拥抱angular2的趋势


How to use(for v2.x)
-------
foo.js
```js
define('foo', ['base'], function(base) {
    base("foo", {
        init: function(initParam) {
            return this;
        },
        sayHello: function(name) {
            return 'hi- ' + name;
        },
        foo: function() {
            return 'foo!';
        },
        end: 0
    });
    base.foo.defaults = {
        end: 0
    };
    return base.foo;
});
```

bar.js, 继承于foo.js
```js
define('bar', function(require) {
    var base = require('base');
    require('foo');
    base("foo.bar", {
        init: function(initParam) {
            return this;
        },
        sayHello: function(name) {
            var superName = this._super('sayHello', name);
            return 'call super:' + superName + ', call self: hello- '
                        + name + "/foo:" + this.foo();
        },
        end: 0
    });
    base.bar.defaults = {
        end: 0
    };
    return base.bar;
});
```

index.html
```js
<!DOCTYPE html>
<html>
<head lang="en">
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Hello perfmjs</title>
    <script type="text/javascript" src="build/core.js?v=20140813001"></script>
    <script>
    require(['loader'], function(loader) {
        loader.loadModules({mdCallback:function(source, module, combineUrls) {
            combineUrls[combineUrls.length] = 'foo.js';
            combineUrls[combineUrls.length] = 'bar.js';
        }, afterLoadedCallback:function() {
            //应用入口函数
            require(['app', 'bar'], function (app, bar) {
                app.registerAndStart(bar);
                alert(bar.instance.sayHello('bar'));
            });
        }});
    )};
    </script>
</head>
<body>
Hello perfmjs!
</body>
</html>
```

License
-------

Copyright 2011 Tony

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

History.md
3.0.6-rc6/ 2015-08-28
==================
* improve: bump to 3.0.6-rc5
* fix: fix bugs in Safari

3.0.6-rc4/ 2015-08-26
==================
* improve: improve login and jwt
* add: add popup

3.0.6-rc3/ 2015-08-24
==================
* add: bump to 3.0.6-rc3
* improve: in gulpfile.js: refactor angular2 directives to a common file

3.0.6-rc2/ 2015-08-20
==================
* add: add RouteData sample to angular2-try
* release: release perfmjs3.0.6
* add: add fetch polyfill

3.0.5/ 2015-08-20
==================
* improve: bump to angular2.0.0-alpha.35

3.0.5-rc3/ 2015-08-19
==================
* add: add 'common-login' angular2 directives
* add: add window.fetch polyfill in perfmjs/utils
* improve: some improvement
* add: add cmsEdit

3.0.4/ 2015-08-12
==================
* improve: improve perfmjs3 to 3.0.4

3.0.3/ 2015-08-08
==================
* improve: angular2-try提升angular2的依赖from 2.0.0-alpha.33 to 2.0.0-alpha.34

3.0.2/ 2015-08-05
==================
* add: angular2-try增加Modal指令,MessageEvent功能

3.0.1/ 2015-08-02
==================
* add: gulp.js增加perfmjs3的build，使用`gulp perfmjs3`
* add: angular2-try增加jsonPipe, filterPipe的功能
* add: angular2-try完善投注功能

3.0.0/ 2015-07-31
==================
* add: perfmjs3.0.0

3.0.0-alpha.1/ 2015-07-29
==================
* init: perfmjs3.0.0-alpha.1, write using TypeScript

2.1.4-rc4/ 2015-07-15
==================
* add: 增加ssq的angular2-try例子

2.1.4-rc3/ 2015-07-15
==================
* improve: 提升angular2的依赖版本到2.0.0-alpha.31

2.1.4-rc2/ 2015-07-10
==================
* improve: 完善angular2-try例子

2.1.3/ 2015-06-23
==================
* fixed: 修复globalConfig的加载BUG
* improve: 新增angular2例子

2.1.2 / 2015-04-20
==================
* improve: 将系统所有公共的插件配置统一放到一个global-config.js文件中管理
* improve: improved jasmine from 2.1.3 to 2.2.1

2.1.1 / 2015-01-20
==================
* fixed: amd.js代码中依赖数组的insert顺序搞反了(splice),这个BUG导致错误地按依赖列表的反序来加载依赖js文件
* improve: 将utils.version改为utils.VERSION

2.1.0 / 2015-01-19
==================
* add: 增加angular.js的例子及测试用例

2.0.0 / 2014-12-2
==================
* improve: 给符合AMD规范的js文件增加版本号功能；优化include资源文件的加载方式

1.3.11 / 2014-10-21
==================
* add: 增加app-config.js

1.3.10 / 2014-09-15
==================
* improve: 将app.js中error.message改为error.stack
* fixed: 修复base.js的option方法的BUG

1.3.9 / 2014-09-02
==================
* improved:简化includeRes#loadModules方法的使用

1.3.8 / 2014-09-01
==================
* fix: 将perfmjs.utils#isH5Supported方法名改成isH5Support()
* improve: 简化核心代码

1.3.7 / 2014-08-28
==================
* improve: amd.js 增加错误处理

1.3.6 / 2014-08-27
==================
* improve: app.js 管理对象生命周期时，去掉moduleId入参
* improve: base.js 增加getName()方法
* improve:.includeres.js 增加对带版本号的（如：core-1.3.6-min.js）这种依赖文件的正则匹配
* improve: 整理命名空间

1.3.5 / 2014-08-26
==================
 * fix: 修复amd.js中不能同步加载资源文件的问题
 * improve: 优化promise/A+实现类async.js
 * add: 增加gulp构建工具文件：gulpfile.js

1.3.4 / 2014-08-22
==================
 * add: 在utils中增加方法：isBrowserSupport()和isNodeJSSupport()
 * add: 增加promise/A+实现类async.js

1.3.3 / 2014-08-19
==================
 * add: 给perfmjs#utils, joquery.js, event.proxy.js, fsm.js, base.js增加AMD模块化规范并在README.md中增加使用例子

1.3.2 / 2014-08-14
==================
 * add: 给includeres.js增加AMD模块化规范并在README.md中增加使用例子

1.3.1 / 2014-08-13
==================
 * add: 给perfmjs框架增加AMD模块化规范并在README.md中增加使用例子

1.3.0 / 2014-08-12
==================
 * add: 实现浏览器端的AMD模块化规范，功能类似require.js,但比require.js更轻量

1.2.5 / 2014-07-28
==================
 * fixed: fixed logic about perfmjs.utils.isH5Support()

1.2.4 / 2014-07-23
==================
 * improve: improved some function in joquery.js

1.2.3 / 2014-07-16
==================
 * fixed: fixed defaults problem in base.js

1.2.2 / 2014-07-16
==================
 * fixed: fixed scope problem in app.js

1.2.1 / 2014-07-09
==================
 * improve: improved README.md and Help document

1.2.0 / 2014-07-09
==================
 * improve: improved base function

1.1.8 / 2014-07-08
==================
 * improve: improved perfmjs.utils#fmtJSONMsg

1.1.7 / 2014-07-03
==================
 * Modify: adjust project fils construct

1.1.6 / 2014-07-02
==================
 * add: perfmjs.utils#isObject, perfmjs.utils#keys

1.1.5 / 2014-06-25
==================
 * add: perfmjs.utils#fastMap, perfmjs.utils#fastReduce

1.1.4 / 2014-06-25
==================
 * fix: fix some bugs
 * add: perfmjs.utils#fastBind, perfmjs.utils#forEach
 * modify: joquery#where, joquery#updateOrInsert, joquery#last, joquery#last

1.1.3 / 2014-06-20
==================
 * fix: fix some bugs
 * add: perfmjs.utils#toNumber, perfmjs.utils#toBoolean

说明：修改类型为: init(初始化项目)/improve(改善功能)/add(新增功能)/remove(移除功能)/modify(修复功能内容)/fix(修改BUG)/release(发布)

js/src/load.js
/*
 * HeadJS     The only script in your <HEAD>    
 * Author     Tero Piirainen  (tipiirai)
 * Maintainer Robert Hoffmann (itechnology)
 * License    MIT / http://bit.ly/mit-license
 *
 * Version 0.99
 * http://headjs.com
 */
//for perfmjs: 去掉注释中的!,并将以下的;改为!o
!(function (win, undefined) {
    "use strict";

    var doc = win.document,
        domWaiters = [],
        queue      = [], // waiters for the "head ready" event
        handlers   = {}, // user functions waiting for events
        assets     = {}, // loadable items in various states
        isAsync    = "async" in doc.createElement("script") || "MozAppearance" in doc.documentElement.style || win.opera,
        isHeadReady,
        isDomReady,

        /*** public API ***/
        headVar = win.head_conf && win.head_conf.head || "head",
        api     = win[headVar] = (win[headVar] || function () { api.ready.apply(null, arguments); }),

        // states
        PRELOADING = 1,
        PRELOADED  = 2,
        LOADING    = 3,
        LOADED     = 4;

    // Method 1: simply load and let browser take care of ordering
    if (isAsync) {
        api.load = function () {
            ///<summary>
            /// INFO: use cases
            ///    head.load("http://domain.com/file.js","http://domain.com/file.js", callBack)
            ///    head.load({ label1: "http://domain.com/file.js" }, { label2: "http://domain.com/file.js" }, callBack)
            ///</summary> 
            var args      = arguments,
                 callback = args[args.length - 1],
                 items    = {};

            if (!isFunction(callback)) {
                callback = null;
            }

            each(args, function (item, i) {
            	if (perfmjs.utils.isArray(item)) {
            		//for perfmjs: 处理第一个参数是资源文件地址数组的情况
            		var item_els = item;
	        		var item_len = item.length;
	        		for (var i = 0; i < item_len; i++) {
	        			item             = getAsset(item_els[i]);
	                    items[item.name] = item;
	                    load(item, callback && i === args.length - 2 ? function () {
	                        if (allLoaded(items)) {
	                            one(callback);
	                        }
	                    } : null);
	        		}
            	} else if (item !== callback) {
                    item             = getAsset(item);
                    items[item.name] = item;

                    load(item, callback && i === args.length - 2 ? function () {
                        if (allLoaded(items)) {
                            one(callback);
                        }

                    } : null);
                }
            });

            return api;
        };


    // Method 2: preload with text/cache hack
    } else {
        api.load = function () {
            var args = arguments,
                rest = [].slice.call(args, 1),
                next = rest[0];

            // wait for a while. immediate execution causes some browsers to ignore caching
            if (!isHeadReady) {
                queue.push(function () {
                    api.load.apply(null, args);
                });

                return api;
            }            

            // multiple arguments
            if (!!next) {
                /* Preload with text/cache hack (not good!)
                 * http://blog.getify.com/on-script-loaders/
                 * http://www.nczonline.net/blog/2010/12/21/thoughts-on-script-loaders/
                 * If caching is not configured correctly on the server, then items could load twice !
                 *************************************************************************************/
                each(rest, function (item) {
                    if (!isFunction(item)) {
                        preLoad(getAsset(item));
                    }
                });

                // execute
                load(getAsset(args[0]), isFunction(next) ? next : function () {
                    api.load.apply(null, rest);
                });                
            }
            else {
            	if (perfmjs.utils.isArray(args[0])) {
                	//for perfmjs: 处理第一个参数是资源文件地址数组的情况
            		var item_els = args[0];
            		var item_len = item_els.length;
            		for (var i = 0; i < item_len; i++) {
            			load(getAsset(item_els[i]));
            		}
            	} else {
	                // single item
	                load(getAsset(args[0]));
            	}
            }

            return api;
        };
    }

    // INFO: for retro compatibility
    api.js = api.load;
    
    api.test = function (test, success, failure, callback) {
        ///<summary>
        /// INFO: use cases:
        ///    head.test(condition, null       , "file.NOk" , callback);
        ///    head.test(condition, "fileOk.js", null       , callback);        
        ///    head.test(condition, "fileOk.js", "file.NOk" , callback);
        ///    head.test(condition, "fileOk.js", ["file.NOk", "file.NOk"], callback);
        ///    head.test({
        ///               test    : condition,
        ///               success : [{ label1: "file1Ok.js"  }, { label2: "file2Ok.js" }],
        ///               failure : [{ label1: "file1NOk.js" }, { label2: "file2NOk.js" }],
        ///               callback: callback
        ///    );  
        ///    head.test({
        ///               test    : condition,
        ///               success : ["file1Ok.js" , "file2Ok.js"],
        ///               failure : ["file1NOk.js", "file2NOk.js"],
        ///               callback: callback
        ///    );         
        ///</summary>    
        var obj = (typeof test === 'object') ? test : {
            test: test,
            success: !!success ? isArray(success) ? success : [success] : false,
            failure: !!failure ? isArray(failure) ? failure : [failure] : false,
            callback: callback || noop
        };

        // Test Passed ?
        var passed = !!obj.test;

        // Do we have a success case
        if (passed && !!obj.success) {
            obj.success.push(obj.callback);
            api.load.apply(null, obj.success);
        }
            // Do we have a fail case
        else if (!passed && !!obj.failure) {
            obj.failure.push(obj.callback);
            api.load.apply(null, obj.failure);
        }
        else {
            callback();
        }

        return api;
    };

    api.ready = function (key, callback) {
        ///<summary>
        /// INFO: use cases:
        ///    head.ready(callBack)
        ///    head.ready(document , callBack)
        ///    head.ready("file.js", callBack);
        ///    head.ready("label"  , callBack);        
        ///</summary>

        // DOM ready check: head.ready(document, function() { });
        if (key === doc) {
            if (isDomReady) {
                one(callback);
            }
            else {
                domWaiters.push(callback);
            }

            return api;
        }

        // shift arguments
        if (isFunction(key)) {
            callback = key;
            key      = "ALL";
        }

        // make sure arguments are sane
        if (typeof key !== 'string' || !isFunction(callback)) {
            return api;
        }

        // This can also be called when we trigger events based on filenames & labels
        var asset = assets[key];

        // item already loaded --> execute and return
        if (asset && asset.state === LOADED || key === 'ALL' && allLoaded() && isDomReady) {
            one(callback);
            return api;
        }

        var arr = handlers[key];
        if (!arr) {
            arr = handlers[key] = [callback];
        }
        else {
            arr.push(callback);
        }

        return api;
    };


    // perform this when DOM is ready
    api.ready(doc, function () {

        if (allLoaded()) {
            each(handlers.ALL, function (callback) {
                one(callback);
            });
        }

        if (api.feature) {
            api.feature("domloaded", true);
        }
    });


    /* private functions
    *********************/
    function noop() {
        // does nothing
    }

    function each(arr, callback) {
        if (!arr) {
            return;
        }

        // arguments special type
        if (typeof arr === 'object') {
            arr = [].slice.call(arr);
        }

        // do the job
        for (var i = 0, l = arr.length; i < l; i++) {
            callback.call(arr, arr[i], i);
        }
    }

    /* A must read: http://bonsaiden.github.com/JavaScript-Garden
     ************************************************************/
    function is(type, obj) {
        var clas = Object.prototype.toString.call(obj).slice(8, -1);
        return obj !== undefined && obj !== null && clas === type;
    }

    function isFunction(item) {
        return is("Function", item);
    }

    function isArray(item) {
        return is("Array", item);
    }

    function toLabel(url) {
        ///<summary>Converts a url to a file label</summary>
        var items = url.split("/"),
             name = items[items.length - 1],
             i    = name.indexOf("?");

        return i !== -1 ? name.substring(0, i) : name;
    }

    // INFO: this look like a "im triggering callbacks all over the place, but only wanna run it one time function" ..should try to make everything work without it if possible
    // INFO: Even better. Look into promises/defered's like jQuery is doing
    function one(callback) {
        ///<summary>Execute a callback only once</summary>
        callback = callback || noop;

        if (callback._done) {
            return;
        }

        callback();
        callback._done = 1;
    }

    function getAsset(item) {
        ///<summary>
        /// Assets are in the form of
        /// { 
        ///     name : label,
        ///     url  : url,
        ///     state: state
        /// }
        ///</summary>
        var asset = {};

        if (typeof item === 'object') {
            for (var label in item) {
                if (!!item[label]) {
                    asset = {
                        name: label,
                        url : item[label]
                    };
                }
            }
        }
        else {
            asset = {
                name: toLabel(item),
                url : item
            };
        }

        // is the item already existant
        var existing = assets[asset.name];
        if (existing && existing.url === asset.url) {
            return existing;
        }

        assets[asset.name] = asset;
        return asset;
    }

    function allLoaded(items) {
        items = items || assets;

        for (var name in items) {
            if (items.hasOwnProperty(name) && items[name].state !== LOADED) {
                return false;
            }
        }
        
        return true;
    }


    function onPreload(asset) {
        asset.state = PRELOADED;

        each(asset.onpreload, function (afterPreload) {
            afterPreload.call();
        });
    }

    function preLoad(asset, callback) {
        if (asset.state === undefined) {

            asset.state     = PRELOADING;
            asset.onpreload = [];

            loadAsset({ url: asset.url, type: 'cache' }, function () {
                onPreload(asset);
            });
        }
    }

    function load(asset, callback) {
        ///<summary>Used with normal loading logic</summary>
        callback = callback || noop;

        if (asset.state === LOADED) {
            callback();
            return;
        }

        // INFO: why would we trigger a ready event when its not really loaded yet ?
        if (asset.state === LOADING) {
            api.ready(asset.name, callback);
            return;
        }

        if (asset.state === PRELOADING) {
            asset.onpreload.push(function () {
                load(asset, callback);
            });
            return;
        }

        asset.state = LOADING;
        
        loadAsset(asset, function () {
            asset.state = LOADED;
            callback();

            // handlers for this asset
            each(handlers[asset.name], function (fn) {
                one(fn);
            });

            // dom is ready & no assets are queued for loading
            // INFO: shouldn't we be doing the same test above ?
            if (isDomReady && allLoaded()) {
                each(handlers.ALL, function (fn) {
                    one(fn);
                });
            }
        });
    }

    /* Parts inspired from: https://github.com/cujojs/curl
    ******************************************************/
    function loadAsset(asset, callback) {
        callback = callback || noop;

        var ele;
        if (/\.css[^\.]*$/.test(asset.url)) {
            ele      = doc.createElement('link');
            ele.type = 'text/' + (asset.type || 'css');
            ele.rel  = 'stylesheet';
            ele.href = asset.url;
        }
        else {
            ele      = doc.createElement('script');
            ele.type = 'text/' + (asset.type || 'javascript');
            ele.src  = asset.url;
        }

        ele.onload  = ele.onreadystatechange = process;
        ele.onerror = error;

        /* Good read, but doesn't give much hope !
         * http://blog.getify.com/on-script-loaders/
         * http://www.nczonline.net/blog/2010/12/21/thoughts-on-script-loaders/
         * https://hacks.mozilla.org/2009/06/defer/
         */

        // ASYNC: load in parellel and execute as soon as possible
        ele.async = false;
        // DEFER: load in parallel but maintain execution order
        ele.defer = false;

        function error(event) {
            event = event || win.event;
            
            // need some more detailed error handling here

            // release event listeners
            ele.onload = ele.onreadystatechange = ele.onerror = null;
                        
            // do callback
            callback();
        }

        function process(event) {
            event = event || win.event;

            // IE 7/8 (2 events on 1st load)
            // 1) event.type = readystatechange, s.readyState = loading
            // 2) event.type = readystatechange, s.readyState = loaded

            // IE 7/8 (1 event on reload)
            // 1) event.type = readystatechange, s.readyState = complete 

            // event.type === 'readystatechange' && /loaded|complete/.test(s.readyState)

            // IE 9 (3 events on 1st load)
            // 1) event.type = readystatechange, s.readyState = loading
            // 2) event.type = readystatechange, s.readyState = loaded
            // 3) event.type = load            , s.readyState = loaded

            // IE 9 (2 events on reload)
            // 1) event.type = readystatechange, s.readyState = complete 
            // 2) event.type = load            , s.readyState = complete 

            // event.type === 'load'             && /loaded|complete/.test(s.readyState)
            // event.type === 'readystatechange' && /loaded|complete/.test(s.readyState)

            // IE 10 (3 events on 1st load)
            // 1) event.type = readystatechange, s.readyState = loading
            // 2) event.type = load            , s.readyState = complete
            // 3) event.type = readystatechange, s.readyState = loaded

            // IE 10 (3 events on reload)
            // 1) event.type = readystatechange, s.readyState = loaded
            // 2) event.type = load            , s.readyState = complete
            // 3) event.type = readystatechange, s.readyState = complete 

            // event.type === 'load'             && /loaded|complete/.test(s.readyState)
            // event.type === 'readystatechange' && /complete/.test(s.readyState)

            // Other Browsers (1 event on 1st load)
            // 1) event.type = load, s.readyState = undefined

            // Other Browsers (1 event on reload)
            // 1) event.type = load, s.readyState = undefined            

            // event.type == 'load' && s.readyState = undefined


            // !doc.documentMode is for IE6/7, IE8+ have documentMode
            if (event.type === 'load' || (/loaded|complete/.test(ele.readyState) && (!doc.documentMode || doc.documentMode < 9))) {
                // release event listeners               
                ele.onload = ele.onreadystatechange = ele.onerror = null;

                // do callback
                callback();
            }

            // emulates error on browsers that don't create an exception
            // INFO: timeout not clearing ..why ?
            //asset.timeout = win.setTimeout(function () {
            //    error({ type: "timeout" });
            //}, 3000);
        }

        // use insertBefore to keep IE from throwing Operation Aborted (thx Bryan Forbes!)
        var head = doc.head || doc.getElementsByTagName('head')[0];
        // but insert at end of head, because otherwise if it is a stylesheet, it will not ovverride values
        head.insertBefore(ele, head.lastChild);
    }

    /* Mix of stuff from jQuery & IEContentLoaded
     * http://dev.w3.org/html5/spec/the-end.html#the-end
     ***************************************************/
    function domReady() {
        // Make sure body exists, at least, in case IE gets a little overzealous (jQuery ticket #5443).
        if (!doc.body) {
            // let's not get nasty by setting a timeout too small.. (loop mania guaranteed if assets are queued)
            win.clearTimeout(api.readyTimeout);
            api.readyTimeout = win.setTimeout(domReady, 50);
            return;
        }

        if (!isDomReady) {
            isDomReady = true;
            each(domWaiters, function (fn) {
                one(fn);
            });
        }
    }

    function domContentLoaded() {
        // W3C
        if (doc.addEventListener) {
            doc.removeEventListener("DOMContentLoaded", domContentLoaded, false);
            domReady();
        }

        // IE
        else if (doc.readyState === "complete") {
            // we're here because readyState === "complete" in oldIE
            // which is good enough for us to call the dom ready!            
            doc.detachEvent("onreadystatechange", domContentLoaded);
            domReady();
        }
    }

    // Catch cases where ready() is called after the browser event has already occurred.
    // we once tried to use readyState "interactive" here, but it caused issues like the one
    // discovered by ChrisS here: http://bugs.jquery.com/ticket/12282#comment:15    
    if (doc.readyState === "complete") {
        domReady();
    }

    // W3C
    else if (doc.addEventListener) {
        doc.addEventListener("DOMContentLoaded", domContentLoaded, false);

        // A fallback to window.onload, that will always work
        win.addEventListener("load", domReady, false);
    }

    // IE
    else {
        // Ensure firing before onload, maybe late but safe also for iframes
        doc.attachEvent("onreadystatechange", domContentLoaded);

        // A fallback to window.onload, that will always work
        win.attachEvent("onload", domReady);

        // If IE and not a frame
        // continually check to see if the document is ready
        var top = false;

        try {
            top = win.frameElement == null && doc.documentElement;
        } catch (e) { }

        if (top && top.doScroll) {
            (function doScrollCheck() {
                if (!isDomReady) {
                    try {
                        // Use the trick by Diego Perini
                        // http://javascript.nwbox.com/IEContentLoaded/
                        top.doScroll("left");
                    } catch (error) {
                        // let's not get nasty by setting a timeout too small.. (loop mania guaranteed if assets are queued)
                        win.clearTimeout(api.readyTimeout);
                        api.readyTimeout = win.setTimeout(doScrollCheck, 50);
                        return;
                    }

                    // and execute any waiting functions
                    domReady();
                }
            })();
        }
    }

    /*for perfmjs begin*/
    perfmjs.loadres = api;
    perfmjs.ready = perfmjs.loadres.ready;
    /*for perfmjs end*/
    
    /*
        We wait for 300 ms before asset loading starts. for some reason this is needed
        to make sure assets are cached. Not sure why this happens yet. A case study:

        https://github.com/headjs/headjs/issues/closed#issue/83
    */
    setTimeout(function () {
        isHeadReady = true;
        each(queue, function (fn) {
            fn();
        });

    }, 300);
})(window);